jaeger-core/src/main/java/com/uber/jaeger/samplers/PerOperationSampler.java

```diff
@@ -0,0 +1,126 @@
+  private final HashMap<String, GuaranteedThroughputSampler> operationNameToSampler;
+  private final int maxOperations;
+  private ProbabilisticSampler defaultSampler;
+  private double lowerBound;
+  private double defaultSamplingProbability;
+
+  public PerOperationSampler(int maxOperations, OperationSamplingParameters strategies) {
+    this(maxOperations,
+         new HashMap<String, GuaranteedThroughputSampler>(),
+         new ProbabilisticSampler(strategies.getDefaultSamplingProbability()));
+    update(strategies);
+  }
+
+  // Visible for testing
+  PerOperationSampler(int maxOperations,
+                      HashMap<String, GuaranteedThroughputSampler> operationNameToSampler,
+                      ProbabilisticSampler defaultSampler) {
+
+    this.operationNameToSampler = operationNameToSampler;
+    this.maxOperations = maxOperations;
+    this.defaultSampler = defaultSampler;
+  }
+
+  /**
+   * Updates the probabilistic samplers for each operation
+   * @param strategies The parameters for operation sampling
+   * @return true iff any samplers were updated
+   */
+  public synchronized boolean update(OperationSamplingParameters strategies) {
+    boolean isUpdated = false;
+
+    lowerBound = strategies.getDefaultLowerBoundTracesPerSecond();
+    defaultSamplingProbability = strategies.getDefaultSamplingProbability();
+    ProbabilisticSampler defaultSampler = new ProbabilisticSampler(strategies.getDefaultSamplingProbability());
+
+    if (!defaultSampler.equals(this.defaultSampler)) {
+      this.defaultSampler = defaultSampler;
+      isUpdated = true;
+    }
+
+    for (PerOperationSamplingParameters strategy : strategies.getPerOperationStrategies()) {
+      String operation = strategy.getOperation();
+      double samplingRate = strategy.getProbabilisticSampling().getSamplingRate();
+      GuaranteedThroughputSampler sampler = new GuaranteedThroughputSampler(samplingRate, lowerBound);
+      GuaranteedThroughputSampler oldSampler = operationNameToSampler.get(operation);
+
+      if (oldSampler == null) {
+        if (operationNameToSampler.size() < maxOperations) {
+          operationNameToSampler.put(operation, sampler);
+          isUpdated = true;
+        } else {
+          log.info("Exceeded the maximum number of operations({}) for per operations sampling",
+                   maxOperations);
+        }
+      } else if (!sampler.equals(oldSampler)) {
+        operationNameToSampler.put(operation, sampler);
+        isUpdated = true;
+      }
+    }
+    return isUpdated;
+  }
+
+  @Override
+  public synchronized SamplingStatus sample(String operation, long id) {
+    GuaranteedThroughputSampler sampler = operationNameToSampler.get(operation);
+    if (sampler != null) {
+      return sampler.sample(operation, id);
+    }
+
+    if (operationNameToSampler.size() < maxOperations) {
+      sampler = new GuaranteedThroughputSampler(lowerBound, defaultSamplingProbability);
```

Comment:
per comment above, if we make default sampler GuaranteedThroughputSampler, then here we can just do something like `sampler = defaultSampler.copy()`. This will make this class simpler.